### ISSUE TITLE
ConnectTimeOut

### ISSUE THREAD

**Issue by notvixer** [sonic182/aiosonic#377]
when sending 999 req, it gets stuck for a while (15-20 sec) then throws a connection error.

code:

```
async def make_req(ses, dx):
   resp = await ses.post(url, json=dx)
   if resp.status_code in (200, 201):
      print("sent data successfully")
   elif resp.status_code == 429:
      await asyncio.sleep(await resp.json()['retry_after'])
```

```
async def main():
   async with aiosonic.HTTPClient() as cli:
      await asyncio.gather(*[asyncio. create_task(make_req(cli, ...)) for i in range(999])
```

error:
```
Traceback (most recent call last):
  File "/data/data/com.termux/files/usr/lib/python3.11/asyncio/tasks.py", line 490, in wait_for
    return fut.result()
           ^^^^^^^^^^^^
  File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/aiosonic/connection.py", line 48, in connect
    await self._connect(urlparsed, verify, ssl_context, dns_info, http2)
  File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/aiosonic/connection.py", line 95, in _connect
    self.reader, self.writer = await open_connection(                                                                               ^^^^^^^^^^^^^^^^^^^^^^
  File "/data/data/com.termux/files/usr/lib/python3.11/asyncio/streams.py", line 48, in open_connection
    transport, _ = await loop.create_connection(
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/data/com.termux/files/usr/lib/python3.11/asyncio/base_events.py", line 1098, in create_connection
    transport, protocol = await self._create_connection_transport(
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/data/com.termux/files/usr/lib/python3.11/asyncio/base_events.py", line 1131, in _create_connection_transport
    await waiter
asyncio.exceptions.CancelledError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/aiosonic/connectors.py", line 92, in after_acquire
    await wait_for(
  File "/data/data/com.termux/files/usr/lib/python3.11/asyncio/tasks.py", line 492, in wait_for
    raise exceptions.TimeoutError() from exc
TimeoutError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/data/data/com.termux/files/home/spam/main.py", line 36, in <module>
    asyncio.run(main())
  File "/data/data/com.termux/files/usr/lib/python3.11/asyncio/runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/data/data/com.termux/files/usr/lib/python3.11/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/data/com.termux/files/usr/lib/python3.11/asyncio/base_events.py", line 650, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/data/data/com.termux/files/home/spam/main.py", line 32, in main
    async with aiosonic.HTTPClient(TCPConnector(pool_size=999)) as ses:
  File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/aiosonic/__init__.py", line 508, in __aexit__
    raise exc
  File "/data/data/com.termux/files/home/spam/main.py", line 33, in main
    await asyncio.gather(*[asyncio.create_task(make_req(ses, ...)) for i in range(999)])
  File "/data/data/com.termux/files/home/spam/main.py", line 21, in make_req
    resp = await ses.post(url, json=data)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/aiosonic/__init__.py", line 592, in post
    return await self._request_with_body(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/aiosonic/__init__.py", line 538, in _request_with_body
    return await self.request(
           ^^^^^^^^^^^^^^^^^^^
  File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/aiosonic/__init__.py", line 768, in request
    response = await wait_for(
               ^^^^^^^^^^^^^^^
  File "/data/data/com.termux/files/usr/lib/python3.11/asyncio/tasks.py", line 479, in wait_for
    return fut.result()
           ^^^^^^^^^^^^
  File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/aiosonic/__init__.py", line 418, in _do_request
    async with (await connector.acquire(*args)) as connection:
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/aiosonic/connectors.py", line 74, in acquire
    return await self.after_acquire(
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/aiosonic/connectors.py", line 97, in after_acquire
    raise ConnectTimeout()
aiosonic.exceptions.ConnectTimeout
```

**Comment by sonic182** [comment]
Oh, ok. In your case I think it is the target server that closes the connection after 999 requests over the same connection.

I'm gonna implement a counter for the connections, to make max `N` number of requests over a connection then close it, wheneve I get time.

**Comment by sonic182** [comment]
I have implemented the conn_max_requests in 0.16.0  https://github.com/sonic182/aiosonic/blob/master/CHANGELOG.md#0160-2023-03-26
